src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -796,6 +796,7 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
 
                     let hir_id = self.cx.tcx.hir().as_local_hir_id(local);
                     if !self.cx.tcx.privacy_access_levels(LOCAL_CRATE).is_exported(hir_id)
+                        && (item.visibility == Visibility::Public)
```

Comment:
shouldn't we check the exportedness of the current item? I think this will break for reexports, it will fail to warn on

```rust
pub use foo::Bar;
mod foo {
   /// [`Private`]
   struct Bar;
   struct Private;
}
```

and it will erroneously warn on

```rust
mod foo {
    /// [`Private`]
    pub struct Foo;
    struct Private;
}
```

Reply:
```rust
pub use foo::Bar;
mod foo {
   /// [`Private`]
   struct Bar;
   struct Private;
}
```
This is a compile error anyway, you can't publicly export a private type.

For the second case I'm not sure why that's an erroneous error? If the type is being documented that implies it's reachable. It would be a good test case though.

Reply:
Oh, I guess this is only triggered for types being documented 

But you can absolutely re-export private types, that's what reexports are used for typically. You cannot expose a private type that has no externally-visible path but that's different.

Reply:
Yes, the first case doesn't compile. `struct Bar` needs to be `pub` to be re-exported.

The second case does not actually warn. Neither before this PR nor after. Iirc, I tested this before already, and it seemed to me the specific code path I touched isn't even involved in this case.

Reply:
Right, because the module is private and the type is not re-exported. So the struct isn't documented at all. It's not a bad idea to add a test case there's no warning but otherwise I don't think we need to change anything.

Reply:
Good idea. I'll add a test case.

Reply:
Oh, I see, the item itself still needs to be pub for this to work. Makes sense.

Yes please add test cases

Reply:
Tests added as suggested. The cases should now have pretty high or even complete coverage of the various cases. Issue #74134 actually triggered as well inside the private module, so it's a good thing that that is now covered too.